src/gaussian.cc

```diff
@@ -0,0 +1,185 @@
+        double axrat = sqrt((x - pm)/r_major);
+        r_major = sqrt(r_major);
+        double ang = atan2(2 * _cov_xy, _sigma_x_sq - _sigma_y_sq) / 2;
+        if(degrees) ang *= M_180_PI;
+
+        return std::make_shared<EllipseMajor>(r_major, axrat, ang, degrees);
+    }
+
+    void Covariance::set_sigma_x_sq(double sigma_x_sq) {
+        if(!(sigma_x_sq >= 0))
+        {
+            throw std::invalid_argument("Invalid sigma_x_sq=" + std::to_string(sigma_x_sq) +
+                                        "; sigma_x_sq >= 0 required.");
+        }
+        _sigma_x_sq = sigma_x_sq;
+    }
+
+    void Covariance::set_sigma_y_sq(double sigma_y_sq) {
+        if(!(sigma_y_sq >= 0))
+        {
+            throw std::invalid_argument("Invalid sigma_y_sq=" + std::to_string(sigma_y_sq) +
+                                        "; sigma_y_sq >= 0 required.");
+        }
+        _sigma_y_sq = sigma_y_sq;
+    }
+
+    void Covariance::set_cov_xy(double cov_xy) {
+        // Take individual sqrt just to be safe and avoid potential overflow
+        double offdiag_max = sqrt(_sigma_x_sq) * sqrt(_sigma_y_sq);
+        double rho = cov_xy / offdiag_max;
+        if(!(rho > -1 && rho < 1))
+        {
+            throw std::invalid_argument("Invalid cov_xy=" + std::to_string(cov_xy) + " with implied rho="
+                                        + std::to_string(rho) + "; -1 < rho < 1 required.");
+        }
+        _cov_xy = cov_xy;
+    }
+
+    void Ellipse::convolve(const Ellipse& ell)
+    {
+        double sigma_x_ell = ell.get_sigma_x();
+        double sigma_y_ell = ell.get_sigma_y();
+        double sigma_x = sqrt(_sigma_x*_sigma_x + sigma_x_ell*sigma_x_ell);
+        double sigma_y = sqrt(_sigma_y*_sigma_y + sigma_y_ell*sigma_y_ell);
+        double rho = (this->get_cov_xy() + ell.get_cov_xy()) / (sigma_x * sigma_y);
+        this->set(sigma_x, sigma_y, rho);
+    }
+
+    EllipseTerms Ellipse::get() const
+    {
+        return EllipseTerms(_sigma_x, _sigma_y, _rho);
+    };
+
+    std::shared_ptr<Ellipse> Ellipse::make_convolution(const Ellipse& ell) const
+    {
+        std::shared_ptr<Ellipse> ell_ret = std::make_shared<Ellipse>(_sigma_x, _sigma_y, _rho);
+        ell_ret->convolve(ell);
+        return std::move(ell_ret);
+    }
+
+    std::shared_ptr<EllipseMajor> Ellipse::make_ellipse_major(bool degrees) const
+    {
+        return std::move(Covariance(*this).make_ellipse_major(degrees=degrees));
+    }
+
+    void Ellipse::set(double sigma_x, double sigma_y, double rho) {
+        Ellipse::check(sigma_x, sigma_y, rho);
+        _rho = rho;
+        _sigma_x = sigma_x;
+        _sigma_y = sigma_y;
+    }
+
+    void Ellipse::set_rho(double rho) {
+        if(!(rho > -1 && rho < 1))
+        {
+            throw std::invalid_argument("Invalid rho=" + std::to_string(rho) + "; -1 < rho < 1 required.");
+        }
+        _rho = rho;
+    }
+
+    void Ellipse::set_sigma_x(double sigma_x) {
+        if(!(sigma_x >= 0))
```

Comment:
`sigma_x > 0` and similary `sigma_y` below

Reply:
I actually specifically want to support either or both being sigmas being zero (both would be a point source).